iothub/device/src/Fx/Fx.cs

```diff
@@ -2,53 +2,48 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
-using System.Collections.ObjectModel;
+using System.Collections.Generic;
+using System.Linq;
 using System.Reflection;
 using System.Runtime.InteropServices;
 
 namespace Microsoft.Azure.Devices.Client
 {
     internal static class Fx
     {
+        private static readonly HashSet<Type> s_fatalExceptionTypes = new()
+        {
+            typeof(OperationCanceledException),
+            typeof(ObjectDisposedException),
+            typeof(ArgumentException),
+            typeof(InvalidOperationException),
+            typeof(InvalidOperationException),
+            typeof(OutOfMemoryException),
+            typeof(SEHException),
+            typeof(NullReferenceException),
+        };
+
         internal static bool IsFatal(Exception ex)
         {
-            while (ex != null)
+            if (s_fatalExceptionTypes.Any(fatalType => fatalType.IsAssignableFrom(ex.GetType())))
```

Comment:
IsAssignableFrom would include an exception like TaskCanceledException which inherits from OperationCanceledException (which is in the list above).

This is like using `if (ex is OperationCanceledException)`.